File: example/tic-tac-toe/tictactoe-frontend/src/Models/GameModels.ts
export type Mark = "X" | "O";

export interface Coord {
  row: number;
  column: number;
}

export interface Play {
  mark: Mark;
  coord: Coord;
}

export interface GameState {
  turn: Mark;
  grid: Mark[][];
  winner: Mark | undefined;
  draw: boolean;
}

export interface ConnectedOpponent {
  isBot: boolean;
}

export interface ConnectedGame {
  state: GameState;
  mark: Mark;
  opponent: ConnectedOpponent;
}

File: example/tic-tac-toe/TicTacToe.Shared/Models/GameModels.cs
using Orleans;

namespace TicTacToe.Shared.Models;

[GenerateSerializer]
public enum Mark
{
    [Id(0)]
    X,

    [Id(1)]
    O
}

[GenerateSerializer]
public record Coord(int Row, int Column);

[GenerateSerializer]
public record Play(Mark Mark, Coord Coord);

[GenerateSerializer]
public record GameState(Mark Turn, Mark?[][] Grid, Mark? Winner, bool Draw)
{
    public bool GameOver => Draw || Winner != null;
}

[GenerateSerializer]
public record ConnectedPlayer(string Id, Mark Mark, bool IsBot);
